### PR TITLE
Better advantage estimation for OnPolicyAlgorithm

### PR DESCRIPTION
**Pull request by TheFUlDeep**
Closes https://github.com/DLR-RM/stable-baselines3/issues/183

Better advantage estimation for OnPolicyAlgorithm

## Description
* Improve advantage estimation for OnPolicyAlgorithm
* Update compute_returns_and_advantage loop

## Motivation and Context
Closes https://github.com/DLR-RM/stable-baselines3/issues/183

- [x] I have raised an issue to propose this change ([required](https://github.com/DLR-RM/stable-baselines3/blob/master/CONTRIBUTING.md) for new features and bug fixes)

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)

## Checklist:

- [x] I've read the [CONTRIBUTION](https://github.com/DLR-RM/stable-baselines3/blob/master/CONTRIBUTING.md) guide (**required**)
- [ ] I have updated the changelog accordingly (**required**).
- [ ] My change requires a change to the documentation.
- [ ] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [ ] I have updated the documentation accordingly.
- [ ] I have reformatted the code using `make format` (**required**)
- [ ] I have checked the codestyle using `make check-codestyle` and `make lint` (**required**)
- [ ] I have ensured `make pytest` and `make type` both pass. (**required**)

Note: we are using a maximum length of 127 characters per line

### CODE REVIEW

**Comment by TheFUlDeep** [comment]
I don't know why codestyle check failed

**Comment by Miffyli** [comment]
Please fill in the PR template and complete the points listed in it as well.

**Comment by TheFUlDeep** [comment]
I don't want to mess with changelog and so on. You are on your own here.

File: stable_baselines3/common/buffers.py [hunk 1]
```diff
@@ -318,16 +318,15 @@ def compute_returns_and_advantage(self, last_value: th.Tensor, dones: np.ndarray
...
         last_gae_lam = 0
-        for step in reversed(range(self.buffer_size)):
-            if step == self.buffer_size - 1:
-                next_non_terminal = 1.0 - dones
-                next_value = last_value
-            else:
-                next_non_terminal = 1.0 - self.dones[step + 1]
-                next_value = self.values[step + 1]
-            delta = self.rewards[step] + self.gamma * next_value * next_non_terminal - self.values[step]
+        next_non_terminal = 1.0 - dones
```
Comment (araffin) [inline stable_baselines3/common/buffers.py:321]: Please keep the previous structure (with `for step in `)

File: stable_baselines3/common/on_policy_algorithm.py [hunk 2]
```diff
@@ -179,6 +179,11 @@ def collect_rollouts(
             self._last_obs = new_obs
             self._last_dones = dones
 
+        with th.no_grad():
+            # Convert to pytorch tensor
```
Comment (araffin) [inline stable_baselines3/common/on_policy_algorithm.py:183]:
```suggestion
            # Compute value for the last timestep
```

**Comment by araffin** [comment]
>I don't want to mess with changelog and so on. You are on your own here.

What do you mean?

File: stable_baselines3/common/buffers.py (hunk 1 above)
Comment (TheFUlDeep) [reply stable_baselines3/common/buffers.py:321]: Iterators are more pythonic way.
Why do you want it?

**Comment by TheFUlDeep** [comment]
>What do you mean?

The template requires to update changelog and stuff. But Im too lazy for it.

**Comment by araffin** [comment]
>The template requires to update changelog and stuff. But Im too lazy for it.

Alright, thanks for your honesty, I will close this PR then.

**Pull request closed without merging**